platform/include/platform/CircularBuffer.h

```diff
@@ -165,43 +285,63 @@ class CircularBuffer {
         core_util_critical_section_exit();
     }
 
-    /** Get the number of elements currently stored in the circular_buffer */
+    /**
+     * Get the number of elements currently stored in the circular_buffer.
+     */
     CounterType size() const
     {
         core_util_critical_section_enter();
-        CounterType elements;
-        if (!_full) {
-            if (_head < _tail) {
-                elements = BufferSize + _head - _tail;
-            } else {
-                elements = _head - _tail;
-            }
-        } else {
-            elements = BufferSize;
-        }
+        CounterType elements = non_critical_size();
         core_util_critical_section_exit();
         return elements;
     }
 
-    /** Peek into circular buffer without popping
+    /** Peek into circular buffer without popping.
      *
-     * @param data Data to be peeked from the buffer
-     * @return True if the buffer is not empty and data contains a transaction, false otherwise
+     * @param data Data to be peeked from the buffer.
+     * @return True if the buffer is not empty and data contains a transaction, false otherwise.
      */
     bool peek(T &data) const
     {
         bool data_updated = false;
         core_util_critical_section_enter();
         if (!empty()) {
-            data = _pool[_tail];
+            data = _buffer[_tail];
             data_updated = true;
         }
         core_util_critical_section_exit();
         return data_updated;
     }
 
 private:
-    T _pool[BufferSize];
+    bool non_critical_empty() const
+    {
+        bool is_empty = (_head == _tail) && !_full;
+        return is_empty;
+    }
+
+    CounterType non_critical_size() const
+    {
+        CounterType elements;
+        if (!_full) {
+            if (_head < _tail) {
+                elements = BufferSize + _head - _tail;
+            } else {
+                elements = _head - _tail;
+            }
+        } else {
+            elements = BufferSize;
+        }
+        return elements;
+    }
+
+    CounterType incrementCounter(CounterType val)
+    {
+        return (++val) % BufferSize;
```

Comment:
If BufferSize isn't a power of two this could be very inefficient on M0-type devices - using an actual software division routine - which would be bad in an interrupt context. You're better off doing a comparison. You know val can't be greater than BufferSize to start with - the compiler doesn't.

Reply:
Would it make sense to force a power of two for the size?

Reply:
Would be a backwards-incompatible API change.